### PR TITLE
AE-2456: Flap including transition triggers on flap cancel

### PR DESCRIPTION
**Pull request by crowt01j**

### CODE REVIEW

File: analysis_engine/library.py [hunk 1]
```diff
@@ -6779,6 +6779,8 @@ def including_transition(array, steps, hz=1, mode='include'):
     for mid_1, flap, mid_2 in zip(mid_steps[:-1], steps, mid_steps[1:]):
         # Slice the data into bands that are between the midpoint flap values
         bands = slices_and(runs_of_ones(array > mid_1), runs_of_ones(array <= mid_2))
+        if len(bands):
+            threshold = max(abs(np.max([np.ma.mean(array[b]) for b in bands]) - flap), 0.1)
```
Comment (kirku1g) [inline analysis_engine/library.py]: Try changing `np.max` to Python built-in `max` as numpy will construct an array from a Python list which will be ~5x slower. Also, this allows the input to be a generator which is more optimal than constructing a list and afterwards iterating over it, e.g. `max(np.ma.mean(array[b]) for b in bands)`

File: analysis_engine/library.py [hunk 2]
```diff
@@ -6789,12 +6791,53 @@ def including_transition(array, steps, hz=1, mode='include'):
                     output[min(band.stop - 1, len(array) - 1)] = flap
                 continue
 
-            if np.ma.count(partial):
+            if np.ma.count(partial) and np.ma.count(partial) == len(partial):
                 # Unchanged data can be included in our output flap array directly
                 output[band] = partial
+            elif np.ma.count(partial):
```
Comment (kirku1g) [inline analysis_engine/library.py]: Calling `np.ma.count` three times will be a bit slower than necessary. I'd recommend:
```
if not partial.mask.any():
    ...
elif not partial.mask.all():
```
`mask.all()`/`mask.any()` are faster than count and `mask.any()` will break out of scanning the array early if any values are truthy.

File: analysis_engine/library.py [hunk 3]
```diff
@@ -6789,12 +6791,53 @@ def including_transition(array, steps, hz=1, mode='include'):
...
+                    index = index[index != len(temp) - 1]
+                if index.size and len(np.unique(temp[index])) > len(temp[index]) * 0.1:
+                    # More than 10% values are unique, so make changes
+                    test = (abs(temp[index + 1] - temp[index]) > min_change) & (temp[index + 1] > temp[index])
+                    flap_up = steps[step_index + 1 if step_index != len(steps) - 1 else len(steps) - 1]
+                    partial[index[test]] = flap_up
+                    # partial[index[~test]] = np.ma.masked
+                    if (
+                        np.ma.count(partial[index[~test]]) and
+                        max([len(partial[x]) for x in np.ma.clump_unmasked(partial[index[~test]])]) <= 60/hz
```
Comment (kirku1g) [inline analysis_engine/library.py]: The argument to `max` could be a generator comprehension.

File: analysis_engine/library.py [hunk 4]
```diff
@@ -6789,12 +6791,53 @@ def including_transition(array, steps, hz=1, mode='include'):
...
+            if np.ma.count(partial) and np.ma.count(partial) == len(partial):
                 # Unchanged data can be included in our output flap array directly
                 output[band] = partial
+            elif np.ma.count(partial):
+                # If partial is not completely unmasked, find values that go above/below threshold.
+                step_index = steps.index(flap)
+                temp = array[band]
+                min_change = np.max(np.diff(steps)) * 0.01
+                # set values within flap threshold to this flap setting.
+                partial[np.logical_and(temp < flap + threshold, temp > flap - threshold)] = flap
```
Comment (kirku1g) [inline analysis_engine/library.py]: `np.logical_and(x, y)` and `np.logical_or(x, y)` are slightly more concisely written as `x & y` and `x | y` respectively, e.g. `(temp < flap + threshold) |  (temp > flap - threshold)`

**Comment by crowt01j** [comment]
Requested changes have been made.

**Comment by kirku1g** [comment]
Resolved separately by AE-2462.

**Pull request closed without merging**